kernel/objects/capability-spinning/objects/RevokeOperation.cc

```diff
@@ -99,98 +99,92 @@ namespace mythos {
     // if some other revoke or delete clears the flag or changes the cap values
     // all children have been deleted in the mean time and we are done
     entry.setRevoking();
-    entry.unlock();
+    entry.unlock_cap();
     _result = _delete(&entry, rootCap).state();
     _startAsyncDelete(t);
   }
 
   optional<void> RevokeOperation::_delete(CapEntry* root, Cap rootCap)
```

Comment:
objective: find first leaf child of the root and remove it from, repeat from the root until no leafs left, then delete root itself.
A CapEntry may be deleted only if it has no children. Otherwise, these children would point to their ancestor's kernel object after it has been deleted. Thus, when deleting an entry, we have to make sure that no children are concurrently inserted after it. The only possibility would be through `CapEntry::insertAfter()`. Alternatively, at a higher level in the inherit operation.
The steps should be: find first leaf, lock the leaf entry in order to prevent new childs to appear behind it, check that it is still the leaf and nobody concurrently inserted a child behind it, delete the kernel object and capability, remove the entry from the chain. If the locking fails we can abort due to concurrency. If the locking is successfull but a child appeared behind, we release the lock and repeat to find the first leaf by starting from where we are.

Reply:
When traversing the CapEntry chain without any locking, it is mandatory to ensure that we still have the same contents in the CapEntry we are looking at, that is the Capabilities and the prev and next pointers. We just check the capability as follows: The initial caller provides the pointer to the CapEntry and the Capability it expects there. When looking at the successor in the chain, we read our next pointer, then read the successor's capability which becomes our expected value for the successor CapEntry. Whenever we decide to modify anything on a CapEntry, we have to acquire a respective lock flag in that entry and, then, verify that the CapEntry still contains the expected content. 
The acquire lock should directly fail if the CapEntry's links do no longer match with what we expect: The lock flags are in the prev and next pointer of the CapEntry. When traversing the chain, we know the expected predecessor of the CapEntry that we try to lock.